src/serac/physics/solid_residual.hpp

```diff
@@ -0,0 +1,629 @@
+  void setMaterial(const MaterialType& material, std::shared_ptr<QuadratureData<StateType>> qdata = EmptyQData)
+  {
+    setMaterial(DependsOn<>{}, material, qdata);
+  }
+
+  /**
+   * @brief Set the material stress response and mass properties for the physics module
+   *
+   * @tparam MaterialType The solid material type
+   * @tparam StateType the type that contains the internal variables for MaterialType
+   * @param material A material that provides a function to evaluate stress
+   * @param domain Domain for material to be integrated over
+   * @pre material must be a object that can be called with the following arguments:
+   *    1. `MaterialType::State & state` an mutable reference to the internal variables for this quadrature point
+   *    2. `tensor<T,dim,dim> du_dx` the displacement gradient at this quadrature point
+   *    3. `tuple{value, derivative}`, a tuple of values and derivatives for each parameter field
+   *            specified in the `DependsOn<...>` argument.
+   *
+   * @note The actual types of these arguments passed will be `double`, `tensor<double, ... >` or tuples thereof
+   *    when doing direct evaluation. When differentiating with respect to one of the inputs, its stored
+   *    values will change to `dual` numbers rather than `double`. (e.g. `tensor<double,3>` becomes `tensor<dual<...>,
+   * 3>`)
+   *
+   * @param qdata the buffer of material internal variables at each quadrature point
+   *
+   * @pre MaterialType must have a public method `density` which can take FiniteElementState parameter inputs
+   * @pre MaterialType must have a public method 'pkStress' which returns the first Piola-Kirchhoff stress
+   *
+   * @note This method must be called prior to completeSetup()
+   */
+  template <int... active_parameters, typename MaterialType, typename StateType = Empty>
+  void setRateMaterial(DependsOn<active_parameters...>, const MaterialType& material, Domain& domain,
+                       qdata_type<StateType> qdata = EmptyQData)
+  {
+    static_assert(std::is_same_v<StateType, Empty> || std::is_same_v<StateType, typename MaterialType::State>,
+                  "invalid quadrature data provided in setMaterial()");
+    RateMaterialStressFunctor<MaterialType> material_functor(material);
+    residual_->AddDomainIntegral(Dimension<dim>{}, DependsOn<0, 1, 2, active_parameters + NUM_STATE_VARS...>{},
+                                 std::move(material_functor), domain, qdata);
+  }
+
+  /// @overload
+  template <typename MaterialType, typename StateType = Empty>
+  void setRateMaterial(const MaterialType& material, std::shared_ptr<QuadratureData<StateType>> qdata = EmptyQData)
+  {
+    setRateMaterial(DependsOn<>{}, material, qdata);
+  }
+
+  /**
+   * @brief Functor representing a body force integrand.  A functor is necessary instead
+   * of an extended, generic lambda for compatibility with NVCC.
+   */
+  template <typename BodyForceType>
```

Comment:
This is another residual term that might benefit from being its own class.  Otherwise, we might risk bloating `solid_residual` with more and more terms that may or may not be used.  Same with traction, etc. below.